[RISCV] Minimally modify incoming state in transferBefore

lukel97 opened the pull request:
transferBefore currently takes an incoming state and an instruction, computes
the new state needed for the instruction, and then modifies that new state to
be more similar to the incoming state.

This patch reverses the approach by instead taking the incoming state and
modifying only the bits that are demanded by the instruction.

It is currently NFC, but later patches would rework adjustIncoming to bring back the improvements from earlier revisions of this PR.

lukel97 committed:
[RISCV] Minimally modify incoming state in transferBefore

transferBefore currently takes an incoming state and an instruction, computes
the new state needed for the instruction, and then modifies that new state to
be more similar to the incoming state.

This patch reverses the approach by instead taking the incoming state and
modifying only the bits that are demanded by the instruction.

I find this makes things slightly easier to reason about. It also us to use the
x0, x0 form in more places, in particular with vmv.x.s since we're no longer
relying on isScalarInsertInstr, but instead reusing the logic in getDemanded.

I haven't had a chance to check, but hopefully this might also address some of
the regressions seen in #71501.

lukel97 committed:
Update test outside of rvv directory

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1013,73 +1015,76 @@ bool RISCVInsertVSETVLI::needVSETVLI(const MachineInstr &MI,
     return;
 
-  const VSETVLIInfo NewInfo = computeInfoForInstr(MI, TSFlags, MRI);
+  VSETVLIInfo NewInfo = computeInfoForInstr(MI, TSFlags, MRI);
```
Please leave NewInfo and PrevInfo unchanged.  Use Info as your mutable state variable if desired.  Mixing the meaning leads to subtle confusions below.

lukel97 replied:
I gave this an attempt, but couldn't figure out a good way of doing so. Mainly because the SEW/LMUL ratio "touch up" has to happen on NewInfo before we actually mutate Info, since CanUseX0X0Form is dependent on said ratio.

We were still essentially mutating NewInfo before this patch, it was just a bit obfuscated because we swapped it into Info first. Any ideas here?

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1013,73 +1015,76 @@ bool RISCVInsertVSETVLI::needVSETVLI(const MachineInstr &MI,
   // maintain the SEW/LMUL ratio. This allows us to eliminate VL toggles in more
   // places.
-  DemandedFields Demanded = getDemanded(MI, MRI, ST);
   if (!Demanded.LMUL && !Demanded.SEWLMULRatio && Info.isValid() &&
```
Unless I'm missing something, this info check needs to be in terms of NewInfo now.

lukel97 replied:
I think this is an unfortunate clashing of two changes here, I've removed the valid/not unknown checks for NewInfo since it's always guaranteed to be known, and just kept the Info (f.k.a PrevInfo) checks. I will try and move this into a prior NFC

lukel97 replied:
Done in 933dd03386df3c0cfedf4fe6bc984d776f1298ca

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1013,73 +1015,76 @@ bool RISCVInsertVSETVLI::needVSETVLI(const MachineInstr &MI,
-  // the 'vsetvli x0, x0, vtype" variant, so we avoid the transform to
-  // prevent extending live range of an avl register operand.
-  // TODO: We can probably relax this for immediates.
-  if (isScalarInsertInstr(MI) && PrevInfo.isValid() &&
```
A good first step (before this change), might be to rework this code in terms of the demanded fields without otherwise changing structure.  (i.e. this can be driven off only demanding zeroness of VL, not isScalarInsertInst directly.)

lukel97 replied:
Done in 69f64dedb09bf39bd3ad66bcbc2c947b23342a11

preames reviewed (commented):
This seems like a generally good idea, some style clean up needed before landing.

lukel97 committed:
Try to reduce code diff

lukel97 committed:
Rework to keep old logic

lukel97 commented:
I have rebased this upon #73909 and added back an early exit that was responsible for a lot of test diffs. Ideally we should be able to view the diff of the four commits past #73909, but GitHub seems to error when I try this:

<img width="432" alt="image" src="https://github.com/llvm/llvm-project/assets/2488460/c6e51fe2-a825-47ca-a127-e4e9df682750">

Here's a link to the diff instead: https://github.com/lukel97/llvm-project/compare/move-avl-reg-coalescing...lukel97:llvm-project:reverse-transferBefore

lukel97 commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1025,9 +1027,13 @@ bool RISCVInsertVSETVLI::needVSETVLI(const MachineInstr &MI,
-// Given an incoming state reaching MI, modifies that state so that it is minimally
-// compatible with MI.  The resulting state is guaranteed to be semantically legal
-// for MI, but may not be the state requested by MI.
+static VSETVLIInfo adjustIncoming(VSETVLIInfo PrevInfo, VSETVLIInfo NewInfo,
```
This forward definition is solely to improve the diff

preames reviewed (approved):
LGTM
